### PR TITLE
Gh2340 Support relation types with no source + add such a relation for Changelog

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/2340

### CODE REVIEW

File: de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5471440_sys_gh2340_ReferenceTarget_In_RelationType.sql [hunk 1]
```diff
@@ -0,0 +1,72 @@
+-- 2017-09-12T14:39:40.248
+-- URL zum Konzept
+INSERT INTO AD_Element (AD_Client_ID,AD_Element_ID,AD_Org_ID,ColumnName,Created,CreatedBy,EntityType,IsActive,Name,PrintName,Updated,UpdatedBy) VALUES (0,543415,0,'IsReferenceTarget',TO_TIMESTAMP('2017-09-12 14:39:40','YYYY-MM-DD HH24:MI:SS'),100,'D','Y','IsReferenceTarget','IsReferenceTarget',TO_TIMESTAMP('2017-09-12 14:39:40','YYYY-MM-DD HH24:MI:SS'),100)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5471440_sys_gh2340_ReferenceTarget_In_RelationType.sql]: Please add a help test about what this new flag is about.
Also, please mention the github URL

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java [hunk 2]
```diff
@@ -108,10 +124,15 @@ public String toString()
 		final ZoomProviderDestination source = sourceAndTarget.getLeft();
 		final ZoomProviderDestination target = sourceAndTarget.getRight();
 
-		if (!source.matchesAsSource(zoomSource))
+		final boolean isReferenceTarget = target.isReferenceTarget();
+
+		if (!isReferenceTarget)
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: please reduce the nesting by combining those ifs

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java [hunk 3]
```diff
@@ -384,6 +435,12 @@ public int getAD_Window_ID(final boolean isSOTrx)
 
 		public boolean matchesAsSource(final IZoomSource zoomSource)
 		{
+			if(isReferenceTarget())
+			{
+				// the source always matches if the target is ReferenceTarget
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: this comment would be great as part of the method's javadoc

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java [hunk 4]
```diff
@@ -220,7 +246,23 @@ private MQuery mkQuery(final IZoomSource zoomSource, final ZoomProviderDestinati
 		final StringBuilder queryWhereClause = new StringBuilder();
 		final ITableRefInfo refTable = target.getTableRefInfo();
 		final String refTableWhereClause = refTable.getWhereClause();
-		if (!Check.isEmpty(refTableWhereClause))
+
+		if (target.isReferenceTarget)
+		{
+			queryWhereClause
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: AFAIR you once implemented support for multiple and somehow prefixed `AD_Table_ID/Record_ID` pairs.
Please also let this zoom-to feature support them

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java [hunk 5]
```diff
@@ -84,8 +85,22 @@ private RelationTypeZoomProvider(final Builder builder)
...
+			source = new ZoomProviderDestination(
+					builder.getSource_Reference_ID(),
+					builder.getSourceTableRefInfoOrNull(),
+					builder.getSourceRoleDisplayName(),
+					false // builder.isReferenceTarget
+			);
+		}
+		else
+		{
+			source = null;
```
Comment (metas-ts) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: A user of RelationTypeZoomProvider might now run into an NPE when calling `RelationTypeZoomProvider .getTableName()`.
Please avoid such a situation. You could e.g. introduce a dedicated ZoomProviderDestination for reference types.

File: de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5471440_sys_gh2340_ReferenceTarget_In_RelationType.sql (hunk 1 above)
Comment (metas-rc) [reply de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5471440_sys_gh2340_ReferenceTarget_In_RelationType.sql]: Done

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java (hunk 5 above)
Comment (metas-rc) [reply de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: I refactored all this part. Please, review.

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java (hunk 2 above)
Comment (metas-rc) [reply de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: Obsolete after refactoring

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java (hunk 4 above)
Comment (metas-rc) [reply de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: I am sorry but I don't remember anything like this. Can you, please, provide more info, if you remember?

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java (hunk 3 above)
Comment (metas-rc) [reply de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvider.java]: I also improved the java doc. Please, ping me if I lost something.

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvidersFactory.java [hunk 6]
```diff
@@ -244,7 +305,7 @@ public RelationTypeZoomProvider getZoomProviderBySourceTableNameAndInternalName(
 
 			try
 			{
-				final RelationTypeZoomProvider zoomProvider = RelationTypeZoomProvider.builder()
+				final RelationTypeZoomProvider zoomProvider = new RelationTypeZoomProvider.Builder()
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/RelationTypeZoomProvidersFactory.java]: pls use the static builder() method

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/ZoomInfoFactory.java [hunk 7]
```diff
@@ -448,13 +448,16 @@ public ZoomInfo retrieveZoomInfo(final IZoomSource source, final int targetWindo
 				.orElseThrow(() -> new AdempiereException("No zoomInfo found for source=" + source + ", targetWindowId=" + targetWindowId));
 	}
 
-	private List<IZoomProvider> retrieveZoomProviders(final String tableName)
+	private List<IZoomProvider> retrieveZoomProviders(final IZoomSource source)
```
[the remark this answers is not in the document]
Comment (metas-rc) [reply de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/model/ZoomInfoFactory.java]: It is enough. Changed it back.

File: de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/server/rpl/api/impl/ImportHelper.java [hunk 8]
```diff
@@ -894,7 +894,7 @@ else if (DisplayType.Integer == adReferenceId
 
 					// 03534: note that a Button can be 'anything' in the case of C_BPartner.AD_OrgBP_ID, it is a table reference to AD_Org
 					|| DisplayType.Button == adReferenceId && column.getAD_Reference_Value_ID() > 0
-					&& X_AD_Reference.VALIDATIONTYPE_TableValidation.equals(column.getAD_Reference_Value().getValidationType()))
+					&& X_AD_Reference.VALIDATIONTYPE_Tabellenvalidierung.equals(column.getAD_Reference_Value().getValidationType()))
```
[the remark this answers is not in the document]
Comment (metas-rc) [reply de.metas.adempiere.adempiere/base/src/main/java-legacy/org/adempiere/server/rpl/api/impl/ImportHelper.java]: Done

**Pull request merged**